transformer_lens/components/mlp.py

```diff
@@ -62,20 +75,13 @@ def __init__(self, cfg: Union[Dict, HookedTransformerConfig]):
     def forward(
         self, x: Float[torch.Tensor, "batch pos d_model"]
     ) -> Float[torch.Tensor, "batch pos d_model"]:
-        # Technically, all these einsums could be done with a single matmul, but this is more readable.
-        pre_act = self.hook_pre(
-            einsum("batch pos d_model, d_model d_mlp -> batch pos d_mlp", x, self.W_in) + self.b_in
-        )  # [batch, pos, d_mlp]
+        # This is equivalent to (roughly) W_in @ x + b_in. It's important to
+        # use a fused addmm to ensure it matches the Huggingface implementation
+        # exactly.
+        pre_act = addmm(self.b_in, self.W_in, x)  # [batch, pos, d_mlp]
         if self.cfg.act_fn is not None and not self.cfg.act_fn.endswith("_ln"):
             post_act = self.hook_post(self.act_fn(pre_act))  # [batch, pos, d_mlp]
         else:
             mid_act = self.hook_mid(self.act_fn(pre_act))  # [batch, pos, d_mlp]
             post_act = self.hook_post(self.ln(mid_act))
```

Comment:
Could you move this into it's own function, so that we can easy apply a unit test to the block?

Reply:
What exactly would be its own function? Something like this?

```
def pre_act(x: Float[torch.Tensor, "batch pos d_model"]) -> Float[torch.Tensor, "batch pos d_mlp"]:
  return addmm(self.b_in, self.W_in, x)
```

Reply:
I was referring to the whole block that applies the hook points. I was imagining that we could have a function that applies the hook points, and then returns `post_act`. That way our entire forward pass turns into nothing more than a call to those two functions, and we can then more easily create unit tests for those pieces in isolation.

Reply:
TBH I'm having a little bit of trouble interpreting this idea. I like that this PR is currently essentially changing two lines, while this change sounds a little more involved, so maybe it would make sense to do separately?

Reply:
That's fine